src/ddmd/visitor.d

```diff
@@ -83,6 +91,146 @@ public:
     void visit(AST.ThrownExceptionExp e) { visit(cast(AST.Expression)e); }
 }
 
+/* Permissive visitor instantiated with the code generation AST family
+ */
+alias SemanticTimePermissiveVisitor = GenericPermissiveVisitor!ASTCodegen;
+
+// Generic permissive visitor where all the nodes do nothing
+private extern (C++) class GenericPermissiveVisitor(AST) : GenericVisitor!AST
+{
+    alias visit = GenericVisitor!AST.visit;
+
+    override void visit(AST.Dsymbol){}
+    override void visit(AST.Parameter){}
+    override void visit(AST.Statement){}
+    override void visit(AST.Type){}
+    override void visit(AST.Expression){}
+    override void visit(AST.TemplateParameter){}
+    override void visit(AST.Condition){}
+    override void visit(AST.Initializer){}
+}
+
+/* Transitive visitor instantiated with the code generation AST family
+ */
+alias SemanticTimeTransitiveVisitor = GenericTransitiveVisitor!ASTCodegen;
+
+// Generic transitive visitor which implements the AST traversal logic for all AST nodes
+private extern (C++) class GenericTransitiveVisitor(AST) : GenericPermissiveVisitor!AST
+{
+    alias visit = GenericPermissiveVisitor!AST.visit;
+
+    mixin MParseVisitMethods!AST;
+
+    override void visit(AST.PeelStatement s)
+    {
+        if (s.s)
+            s.s.accept(this);
+    }
+
+    override void visit(AST.UnrolledLoopStatement s)
+    {
+        foreach(sx; *s.statements)
```

Comment:
I would recommend adding curly braces to the body of the foreach loop, to avoid any confusion of which code belongs to which body. That's a general recommendation when there are nested statements like this.